src/Control/Effect/NonDet.hs

```diff
@@ -1,61 +1,119 @@
+--
+--   prop> branch None Pure Alt a == (a :: Branch e [] a)
+--   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (None a :: Branch [] a) == applyFun f a
+--   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (Pure a :: Branch [] a) == applyFun g a
+--   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (Alt a b :: Branch [] a) == applyFun2 h a b
+branch :: (e -> a) -> (b -> a) -> (m b -> m b -> a) -> Branch m e b -> a
+branch f _ _ (None a)  = f a
+branch _ f _ (Pure a)  = f a
+branch _ _ f (Alt a b) = f a b
+{-# INLINE branch #-}
+
+-- | Interpret a 'Branch' into an underlying 'Alternative' context.
+runBranch :: Alternative m => (e -> m a) -> Branch m e a -> m a
+runBranch f = branch f pure (<|>)
+{-# INLINE runBranch #-}
+
 
 -- | Run a 'NonDet' effect, collecting all branches’ results into an 'Alternative' functor.
 --
 --   Using '[]' as the 'Alternative' functor will produce all results, while 'Maybe' will return only the first. However, unlike 'runNonDetOnce', this will still enumerate the entire search space before returning, meaning that it will diverge for infinite search spaces, even when using 'Maybe'.
 --
 --   prop> run (runNonDet (pure a)) == [a]
 --   prop> run (runNonDet (pure a)) == Just a
-runNonDet :: (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Applicative m) => Eff (AltC f m) a -> m (f a)
-runNonDet = runAltC . interpret
+runNonDet :: AltC f m a -> m (f a)
+runNonDet = runAltC
 
 newtype AltC f m a = AltC { runAltC :: m (f a) }
+  deriving (Functor)
 
-instance (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Applicative m) => Carrier (NonDet :+: sig) (AltC f m) where
-  ret a = AltC (ret (pure a))
-  eff = AltC . handleSum (eff . handleTraversable runAltC) (\case
-    Empty    -> ret empty
-    Choose k -> liftA2 (<|>) (runAltC (k True)) (runAltC (k False)))
+instance (Applicative f, Applicative m) => Applicative (AltC f m) where
+  pure = AltC . pure . pure
+  AltC f <*> AltC a = AltC (liftA2 (<*>) f a)
 
+instance (Alternative f, Applicative m, Monad f, Traversable f) => Alternative (AltC f m) where
+  empty = AltC (pure empty)
+  l <|> r = AltC (liftA2 (<|>) (runNonDet l) (runNonDet r))
 
--- | Run a 'NonDet' effect, returning the first successful result in an 'Alternative' functor.
---
---   Unlike 'runNonDet', this will terminate immediately upon finding a solution.
---
---   prop> run (runNonDetOnce (asum (map pure (repeat a)))) == [a]
---   prop> run (runNonDetOnce (asum (map pure (repeat a)))) == Just a
-runNonDetOnce :: (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Monad m) => Eff (OnceC f m) a -> m (f a)
-runNonDetOnce = runNonDet . runCull . cull . runOnceC . interpret
+instance (Alternative f, Monad f, Monad m, Traversable f) => Monad (AltC f m) where
+  AltC a >>= f = AltC (a >>= runNonDet . Monoid.getAlt . foldMap (Monoid.Alt . f))
+
+instance (Alternative f, Monad f, MonadFail m, Traversable f) => MonadFail (AltC f m) where
```

Comment:
Can we `derive` this instance?

Reply:
Alas not; the `m (f a)` in `AltC`’s body precludes `GeneralizedNewtypeDeriving`.